### ISSUE TITLE
Provide a way to configure TLS from a Gateway to Backends

### ISSUE THREAD

**Issue by robscott** [kubernetes-sigs/gateway-api#1067]
**What would you like to be added**: 
The ability to configure TLS connections between a Gateway and Backends (also commonly referred to as upstream TLS or re-encryption). At a minimum, this should include the ability to select this protocol for these connections. Additionally it would likely be helpful to be able to configure CA bundle and SAN(s) to be used for validation.

**Why this is needed**:
Previously discussed in #968, this is a widely desired feature that is currently missing from the API. 

Note: This kind of feature will require a GEP in the future. This issue is initially intended just to track discussion around both interest level and potential approaches for this potential addition to the API.

**Comment by mehabhalodiya** [comment]
I need your help. I would like to work on this. May I? 
Thanks!

**Comment by youngnick** [comment]
I think that this issue is waiting on feedback, use cases and so on before we generate a GEP and get started on a design. I think that something that could be very helpful is to review what some specific CRDs do (Istio, Contour, Gloo, and Openshift all have CRDs that I *think* let you specify this). That way we can be confident we've checked what the community is doing and be hopeful that a design will be useful for everyone.

**Comment by hbagdi** [comment]
Writing down some thoughts for the future:
TLS between the gateway and the upstream service is about encryption and identity.
This includes a common trust (CA), SANs for the upstream services, and a client certificate in the gateway (in case the upstream wants to verify the identity of the gateway).
The challenge here is to figure out the ownership of the above information:
- Trust stores are _generally_ (not always) defined at a level above gateway or service. It is not recommended and rare to have a trust-store that is specific to a service. ReferencePolicy at the Gateway level seems to be a good fit here.
- The same could be said about the client certificate of the Gateway, although I'm not sure. A gateway could have multiple identities to different services depending on the service it is talking to. Should we optimize for the common case here?
- The SAN to use for validating upstream service is a tricky one. Does the upstream service owner define those? Or is that something the cluster-admin or service admin defines? I can see use-cases either way. Service admin and route owner can be the same or different person(s).

**Comment by mikemorris** [comment]
/remove-lifecycle stale

@candita has started work on supporting this functionality it https://github.com/kubernetes-sigs/gateway-api/pull/1430

**Comment by shaneutt** [comment]
@candita is currently working on this (thank you Candace!)

/assign @candita

**Comment by tsaarni** [comment]
@shaneutt Adding my arguments from #622 here as well, since it was closed as duplicate.

Use case:


1. They generate their own certificates and want to configure _server certificate_ to their backend service.
2. Next the developer would like to protect the backend service even further, by allowing incoming TLS connections only from the proxy. Symmetrically, application developer wants to enable client authentication, like they did for the server authentication. This would require client certificate to be configurable per backend service, not per Gateway which would be outside of application developer's domain.

Implementing this enhancement has following advantages:

* Allows "self-service" workflow for the application developer, since they do not need to depend on cluster admin to configure client certificate on gateway level
* Cluster admin does not need to manage / coordinate / distribute single CA certificate for the development teams working in the cluster, in separate namespaces. Each team can configure their own client certificates for the proxy and CA certificates to their backend for validating the client cert. 


I know I'm responding to old thread but wanted to add

@hbagdi wrote
> * Trust stores are generally (not always) defined at a level above gateway or service. It is not recommended and rare to have a trust-store that is specific to a service. ReferencePolicy at the Gateway level seems to be a good fit here.
> * The same could be said about the client certificate of the Gateway, although I'm not sure. A gateway could have multiple identities to different services depending on the service it is talking to. Should we optimize for the common case here?

I've had experience from my organization that people ask per service configuration like in the use case I wrote above since it has proven to be complicated to coordinate the credentials on the cluster level. It would imply that gateway would need (a) per service trusted CA certificate to validate the server certificate of the backend server, and (b) per service client certificate for authenticating towards the backend service. 

> * The SAN to use for validating upstream service is a tricky one. Does the upstream service owner define those? Or is that something the cluster-admin or service admin defines? I can see use-cases either way. Service admin and route owner can be the same or different person(s).

Assuming the upstream service has a hostname inside the cluster (the name of the `Service`), then typical hostname validation can be done ([RFC 9110](https://www.rfc-editor.org/rfc/rfc9110#name-https-certificate-verificat), [RFC 2818](https://www.rfc-editor.org/rfc/rfc2818#section-3.1)).
